utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/encode/JsonSchemaOutputValidator.scala

```diff
@@ -0,0 +1,103 @@
+package pl.touk.nussknacker.engine.avro.encode
+import cats.data.Validated.condNel
+import cats.data.{NonEmptyList, Validated, ValidatedNel}
+import cats.implicits._
+import com.typesafe.scalalogging.LazyLogging
+import org.everit.json.schema.{ObjectSchema, Schema}
+import pl.touk.nussknacker.engine.api.NodeId
+import pl.touk.nussknacker.engine.api.typed.typing._
+import pl.touk.nussknacker.engine.json.JsonSchemaTypeDefinitionExtractor
+import pl.touk.nussknacker.engine.util.output._
+
+import scala.language.implicitConversions
+
+private[encode] case class JsonSchemaExpected(schema: Schema) extends OutputValidatorExpected {
+  override def expected: String = JsonSchemaOutputValidatorPrinter.print(schema)
+}
+
+object JsonSchemaOutputValidator {
+  private[encode] val SimplePath = "Data" //todo: check why is it needed, and if we can get rid of it.
+}
+
+class JsonSchemaOutputValidator(validationMode: ValidationMode) extends LazyLogging {
+
+  import JsonSchemaOutputValidator._
+
+  import scala.collection.JavaConverters._
+
+  private val valid = Validated.Valid(())
+
+  /**
+    * Currently we support only basic json-schema configurations. See {@link pl.touk.nussknacker.engine.json.JsonSchemaTypeDefinitionExtractor}
+    */
+  def validateTypingResultToSchema(typingResult: TypingResult, parentSchema: Schema)(implicit nodeId: NodeId): ValidatedNel[OutputValidatorError, Unit] =
+    validateTypingResult(typingResult, parentSchema, None)
+
+  //todo: add support for: unions, enums, nested types, logical types
+  final private def validateTypingResult(typingResult: TypingResult, schema: Schema, path: Option[String]): ValidatedNel[OutputValidatorError, Unit] = {
+    (typingResult, schema) match {
+      case (typingResult: TypedObjectTypingResult, s: ObjectSchema) =>
+        validateRecordSchema(typingResult, s, path)
+      case (_@TypedNull, _) if !schema.isNullable =>
+        invalid(typingResult, schema, path)
+      case (_@TypedNull, _) if schema.isNullable => valid
+      case (_, _) => canBeSubclassOf(typingResult, schema, path)
+    }
+  }
+
+  private def validateRecordSchema(typingResult: TypedObjectTypingResult, schema: ObjectSchema, path: Option[String]): Validated[NonEmptyList[OutputValidatorError], Unit] = {
+    val schemaFields = schema.getPropertySchemas.asScala
+
+    val requiredFieldNames = if (validationMode == ValidationMode.strict) {
```

Comment:
let's move this field closer to requiredFieldsValidation (or whole validation in separate method?) so it's more readable

Reply:
done